### PR TITLE
Move signature implementation to a single function, and catch shape errors

### PR DESCRIPTION
**Pull request by phinate**
I'm playing around with dispatching `signax.signature` in such a way that the correct behaviour happens depending on the shape of the path. 

As an example, the function will still work on paths of shape (a, b ,c) and just (d) too, but both of these may not be the desired behaviour. So, this makes sure that if you feed in a path that is not right, we catch it as an error. 

Also, if the path has a batch dim, this will broadcast the signature over the batch with vmap! `flatten` and `num_chunks` have become arguments to `signature`.

### CODE REVIEW

File: src/signax/signatures.py [hunk 1]
```diff
@@ -49,28 +111,32 @@ def _body(carry, path_inc):
...
-        return [
+        res = [
             jnp.concatenate([first[None, ...], rest], axis=0)
             for first, rest in zip(exp_term, stacked)
         ]
-
-    return carry
+    else:
+        res = carry
+    if flatten:
```
Comment (anh-tong) [inline src/signax/signatures.py:120]: Since the shape of `res` when `stream=True` (shape = `[(path_len-1, dim), (path_len-1, dim, dim]`) is different to when `stream=False` (shape = `[(dim,), (dim, dim)]`), we may use `jax.vmap(utils.flatten)` when `stream=True`.  

We vmap the first dimesion `path_len - 1`

File: src/signax/signatures.py [hunk 2]
```diff
@@ -3,36 +3,98 @@
...
+    if path.ndim == 2:
+        return sig_fun(path)  # regular case
+    if path.ndim == 3:
+        if flatten:
+            return jax.vmap(sig_fun)(
+                path
+            )  # can use vmap when flattening since it's all arrays
+        return [
+            sig_fun(path[i]) for i in range(path.shape[0])
+        ]  # otherwise, use scan to handle the batch dimension in the list
```
Comment (anh-tong) [inline src/signax/signatures.py:72]: I think this is not necessary after we handle `flatten` correctly in the `_signature` function.

File: src/signax/signatures.py [hunk 3]
```diff
@@ -142,16 +208,28 @@ def scan_fn(last_sig, current_stream_signature):
...
     # no remainder, just return the bulk
+    if flatten:
+        bulk_signature = utils.flatten(bulk_signature)
     return bulk_signature
 
 
+@partial(jax.jit, static_argnames=["depth", "stream", "flatten", "num_chunks"])
 def logsignature(
-    path: Float[Array, "path_len dim"], depth: int, stream: bool = False
+    path: Float[Array, "path_len dim"],
```
Comment (anh-tong) [inline src/signax/signatures.py:218]: I guess you want to make `logsignature` to handle the input `(batch_size, path_len, dim)` as well. 

In this case, we also need to handle the shape and need an extra `vmap`

File: tests/test_against_signatory.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+
+def test_signature_1d_path():
+    depth = 3
+    length = 100
+    path = rng.standard_normal((length, 1))
+    signature(path, depth)
+
+    path = rng.standard_normal(length)
+    signature(path, depth)
```
Comment (anh-tong) [inline tests/test_against_signatory.py:25]: We can remove this. This actually causes error since `path` is 1D.

**Comment by anh-tong** [comment]
Overall, it looks good. When the behaviour of `flatten` in `_signature` function is corrected and allowing `logsignature` to handle inputs with `(batch_size, path_len, dim)` is added, we can merge this to `main`.

Also, can you please check why CI format has errors? 
image

File: src/signax/signatures.py (hunk 1 above)
Comment (phinate) [reply src/signax/signatures.py:122]: covered this by using `ravel_pytree` and getting rid of `flatten`:

```python3
import jax.numpy as jnp
from jax import flatten_util

arr = [jnp.ones((2, 3)), jnp.ones((2, 3, 4))]
tree = flatten_util.ravel_pytree(arr)[0]
print(tree)
# > [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]
```

File: src/signax/signatures.py (hunk 2 above)
Comment (phinate) [reply src/signax/signatures.py]: so i did this because i couldn't figure out how to get the batch dimension as the dimension of the list! It seems like vmap only correctly broadcasts the batch dimension to the arrays, but maybe I fed things in wrong. This is the only thing I did that gave the correct shape for now.

Comment (phinate) [reply src/signax/signatures.py]: this problem does not exist when `flatten=True`, since it's arrays all the way down!

File: src/signax/signatures.py (hunk 3 above)
Comment (phinate) [reply src/signax/signatures.py]: this should work automatically by dispatching to those checks in `signature`, which should do the right things I think! But let me know if there's logic I'm missing here.

File: tests/test_against_signatory.py (hunk 4 above)
Comment (phinate) [reply tests/test_against_signatory.py]: done!

**Comment by phinate** [comment]
ok thinks look mostly okay now, i think i've caught the majority of errors -- there are still a couple tests failing with `stream=True` -- do you think i'm handling the flattening logic wrong somewhere?

**Comment by anh-tong** [comment]
From the comment
https://github.com/anh-tong/signax/blob/feb0639481f79d5e683f2ee20edc1c2e4ca338c4/src/signax/signatures.py#L72
I'm curious about why we cannot `vmap` a function returning a list.

I test with a simple example like this:
```python
import jax
import jax.numpy as jnp
from jax import flatten_util

def fn(x, flatten=False):
    n = x.shape[0]
    res = [jnp.ones((n, )), jnp.ones((n, n))]
    if flatten:
        res = flatten_util.ravel_pytree(res)[0]
    return res

# normal case
x = jnp.ones((2,))
output = fn(x)
[print(o.shape) for o in output]
# returned shape: [(2,), (2, 2)]

# vmap
print("--- vmap ---")
batch_size = 5
x = jnp.ones((batch_size, 2))
output = jax.vmap(fn)(x)
[print(o.shape) for o in output]
# return [(batch_size, 2), (batch_size, 2, 2)]


# flatten + vmap
print("--- flatten + vmap ---")
batch_size = 5
x = jnp.ones((batch_size, 2))
output = jax.vmap(lambda _x: fn(_x, flatten=True))(x)
print(output.shape)
# return (batch_size, 2 + 2 * 2)
```

We can make `_signature` handle all the arguments (`stream`, `flatten`, ...). The remaining job of `signature` function is just to check the shape and dispatch between `_signature` and vmapped `_signature`

I suggest to handle `flatten` for the case of `stream=True` and `stream=False` in this line:
https://github.com/anh-tong/signax/blob/feb0639481f79d5e683f2ee20edc1c2e4ca338c4/src/signax/signatures.py#L115-L123
with something like this
```python
if stream:
      res = [
          jnp.concatenate([first[None, ...], rest], axis=0)
          for first, rest in zip(exp_term, stacked)
      ]
    # here `res` has shape [(patch_len - 1, dim), (patch_len - 1, dim, dim), ...]
    if flatten:
        res =jax.vmap(lambda _x: flatten_util.ravel_pytree(x)[0])(res)
        # now `res` has shape (patch_len -1, dim + dim * dim + ...)
  else:
      res = carry
      # `res` has shape [(dim,), (dim, dim), ...]
    if flatten:
        res = flatten_util.ravel_pytree(res)[0] # `res` has shape (dim + dim * dim + ..., )
```

I believe the following line works well. For consistency, probably we can change from `signature` function to `_signature` which designates to handle 2D input.

https://github.com/anh-tong/signax/blob/feb0639481f79d5e683f2ee20edc1c2e4ca338c4/src/signax/signatures.py#L214


Also, since I cannot find the tests comparing to `signatory`, I wonder if we can remove it from now on.

**Comment by phinate** [comment]
> I'm curious about why we cannot `vmap` a function returning a list.
> I test with a simple example like this:
> 
> ```python
> import jax
> import jax.numpy as jnp
> from jax import flatten_util
> 
> def fn(x, flatten=False):
>     n = x.shape[0]
>     res = [jnp.ones((n, )), jnp.ones((n, n))]
>     if flatten:
>         res = flatten_util.ravel_pytree(res)[0]
>     return res
> 
> # normal case
> x = jnp.ones((2,))
> output = fn(x)
> [print(o.shape) for o in output]
> # returned shape: [(2,), (2, 2)]
> 
> # vmap
> print("--- vmap ---")
> batch_size = 5
> x = jnp.ones((batch_size, 2))
> output = jax.vmap(fn)(x)
> [print(o.shape) for o in output]
> # return [(batch_size, 2), (batch_size, 2, 2)]
> 
> 
> # flatten + vmap
> print("--- flatten + vmap ---")
> batch_size = 5
> x = jnp.ones((batch_size, 2))
> output = jax.vmap(lambda _x: fn(_x, flatten=True))(x)
> print(output.shape)
> # return (batch_size, 2 + 2 * 2)
> ```
> 

It's the second case that I implemented this line for -- would you not expect `[(2), (2, 2)]*batch_size` instead of `[(batch_size, 2), (batch_size, 2, 2)]` here? That's the hard thing to do with vmap.

> 
> I suggest to handle `flatten` for the case of `stream=True` and `stream=False` in this line:
> 
> https://github.com/anh-tong/signax/blob/feb0639481f79d5e683f2ee20edc1c2e4ca338c4/src/signax/signatures.py#L115-L123
> 
> 
> with something like this
> ```python
> if stream:
>       res = [
>           jnp.concatenate([first[None, ...], rest], axis=0)
>           for first, rest in zip(exp_term, stacked)
>       ]
>     # here `res` has shape [(patch_len - 1, dim), (patch_len - 1, dim, dim), ...]
>     if flatten:
>         res =jax.vmap(lambda _x: flatten_util.ravel_pytree(x)[0])(res)
>         # now `res` has shape (patch_len -1, dim + dim * dim + ...)
>   else:
>       res = carry
>       # `res` has shape [(dim,), (dim, dim), ...]
>     if flatten:
>         res = flatten_util.ravel_pytree(res)[0] # `res` has shape (dim + dim * dim + ..., )
> ```

cool, I'll look at this now -- thanks for your help @anh-tong :D

**Comment by phinate** [comment]
stream fix worked! but the code's a bit messy now -- would need to think about how to handle things for logsignature (i currently force no flattening since `signature_to_logsignature` needs things in list form), but I could probably re-write the flatten logic to happen once in the same way for all functions as opposed to pasting the same code everywhere...

**Comment by anh-tong** [comment]
Hi @phinate, sorry for taking so long. I have updated the code in the following:

- function `_signature`, `_logsignature` just handle the input with shape `(length, channels)`
- function `signature` and `logsignature` uses on top of `_signature` and `_logsignature` just perform shape handling.
- `flatten=True` now is the default option like `signatory`

Indeed, keeping track of the shape can be challenging, especially when considering the stream and flatten options being turned on or off.

Let me know if the behavior of current version is what you expect.

**Review by anh-tong — APPROVED** [review]

**Comment by anh-tong** [comment]
Let's merge this. Thanks for PR.
